### PR TITLE
FEATURE: Add data-tag-groups attribute to tags (Part 1)

### PR DESCRIPTION
**Pull request by Lhcfl**
Overview
========

This PR is part of a new feature that aims to adds the `data-tag-groups`
attribute to tags, making it easier to use CSS selectors to batch select
tags belonging to certain tag groups.

An example:

```html
<a href="/tag/hidden" data-tag-name="hidden"
data-tag-groups="Hidden Tag Group, non-hidden tags"
class="discourse-tag box">hidden</a>
```

This commit implements adding `data-tag-groups` to tags on TopicList

Related meta topic: https://meta.discourse.org/t/add-a-data-tag-group-attribute-for-tags/234410

ScreenShot
========

The following screenshot shows what happens when custom CSS is 
applied to a tag group:
All tags belonging to the test group have a big border-radius

```css
.discourse-tag[data-tag-groups~="test"] {
    border-radius: 999px !important;
}
```

image

DEV Improvements
================

Partially rewrote `defaultRenderTag` for better scalability and clarity:

The new `defaultRenderTag` reserves an interface for theme components to
avoid the need for components to completely rewrite `defaultRenderTag`
to provide custom tag renderer.

For example, you can now replace the tagRenderer like this:

```javascript
api.replaceTagRenderer((tag, params) => {
  return defaultRenderTag(tag, params, {
    extraClass: "custom-tag-class",
    contentFn(old) {
      return `${iconHTML("tag")}${old}`;
    },
  });
});
```

This has a significant benefit for discourse tag icons. Since the
[old way of code](https://github.com/discourse/discourse-tag-icons/blob/main/javascripts/discourse/initializers/tag-icons.js#L9-L84)
the changes made by this PR will be completely overwritten by discourse
tag icons, and the theme component must be updated accordingly.
After applying the new interface, it can automatically synchronize
changes with the `defaultRendererTag` in core.

### CODE REVIEW

**Comment by jjaffeux** [comment]
Please wait for validation from dev-xp on this

File: app/serializers/concerns/topic_tags_mixin.rb [hunk 1]
```diff
@@ -23,6 +24,16 @@ def tags_descriptions
       .compact
   end
 
+  def tags_groups
+    all_tags
+      .each
+      .with_object({}) do |tag, acc|
+        acc[tag.name] = DiscourseTagging.tag_groups_for(tag) &
+          DiscourseTagging.visible_tag_groups_for(scope)
+      end
```
Comment (davidtaylorhq) [inline app/serializers/concerns/topic_tags_mixin.rb:33]: These `DiscourseTagging` calls both involve calls to redis (`Discourse.cache`). So, when this serializer is used to render a topic list, it's potentially going to result in thousands of redis calls.

File: lib/discourse_tagging.rb [hunk 2]
```diff
@@ -794,4 +798,45 @@ def self.muted_tags(user)
...
+        Tag
+          .left_outer_joins(:tag_groups)
+          .select("tags.*, array_agg(tag_groups.name) as tag_group_names")
+          .group("tags.id")
+          .all
+          .each
+          .with_object({}) do |tag, hash|
+            hash[tag.name] = tag.tag_group_names
+            hash[tag.id] = tag.tag_group_names
+          end
```
Comment (davidtaylorhq) [inline lib/discourse_tagging.rb:832]: The number of tags on a site is unlimited. It wouldn't surprise me if we had sites with 100k+. 

So... I don't think we can afford to be holding them all in memory in a single object. It's likely to lead to performance issues on sites with lots of tags.

**Review by davidtaylorhq — COMMENTED** [review]
Given that this affects our two highest-traffic routes (topic, and topic-list), we need to be extremely careful about performance here.

I think my approach would be to look at how tags are preloaded/serialized for topic-lists at the moment, and then apply some kind of `JOIN` or preloading such that adding the tag group info can be done in a single query.

Even if implemented perfectly, this feature will introduce extra costs on these critical routes, so we need to be absolutely sure we want to pay the cost for this feature.

File: lib/discourse_tagging.rb (hunk 2 above)
Comment (Lhcfl) [reply lib/discourse_tagging.rb:832]: Just curious... So what would /tags.json return for a forum with 100K+ tags?

Comment (davidtaylorhq) [reply lib/discourse_tagging.rb:832]: Yeah, we've had problems with that route on some sites, and will need to add pagination.

The good news is, in that case, it's a very rarely hit route. So even if it's slow, it doesn't affect the performance for most visitors.

**Comment by davidtaylorhq** [comment]
I understand we're gonna put a pin in this project for the time being, so closing this out

**Pull request closed without merging**